File: imageedit/src/main/java/com/zhongjh/imageedit/core/clip/ImageClip.java
package com.zhongjh.imageedit.core.clip;

import android.graphics.RectF;

/**
 *
 * @author felix
 * @date 2017/11/28 下午6:15
 */
public interface ImageClip {

    /**
     * 裁剪区域的边距
     */
    float CLIP_MARGIN = 60f;

    /**
     * 角尺寸
     */
    float CLIP_CORNER_SIZE = 48f;

    /**
     * 裁剪区域最小尺寸
     */
    float CLIP_FRAME_MIN = CLIP_CORNER_SIZE * 3.14f;

    /**
     * 内边厚度
     */
    float CLIP_THICKNESS_CELL = 3f;

    /**
     * 外边框厚度
     */
    float CLIP_THICKNESS_FRAME = 8f;

    /**
     * 角边厚度
     */
    float CLIP_THICKNESS_SEWING = 14f;

    /**
     * 比例尺，用于计算出 {0, width, 1/3 width, 2/3 width} & {0, height, 1/3 height, 2/3 height}
     */
    float[] CLIP_SIZE_RATIO = {0, 1, 0.33f, 0.66f};

    int CLIP_CELL_STRIDES = 0x7362DC98;

    int CLIP_CORNER_STRIDES = 0x0AAFF550;

    float[] CLIP_CORNER_STEPS = {0, 3, -3};

    float[] CLIP_CORNER_SIZES = {0, CLIP_CORNER_SIZE, -CLIP_CORNER_SIZE};

    byte[] CLIP_CORNERS = {
            0x8, 0x8, 0x9, 0x8,
            0x6, 0x8, 0x4, 0x8,
            0x4, 0x8, 0x4, 0x1,
            0x4, 0xA, 0x4, 0x8,
            0x4, 0x4, 0x6, 0x4,
            0x9, 0x4, 0x8, 0x4,
            0x8, 0x4, 0x8, 0x6,
            0x8, 0x9, 0x8, 0x8
    };

    enum Anchor {
        /**
         * LEFT
         */
        LEFT(1),
        /**
         * RIGHT
         */
        RIGHT(2),
        /**
         * TOP
         */
        TOP(4),
        /**
         * BOTTOM
         */
        BOTTOM(8),
        /**
         * LEFT_TOP
         */
        LEFT_TOP(5),
        /**
         * RIGHT_TOP
         */
        RIGHT_TOP(6),
        /**
         * LEFT_BOTTOM
         */
        LEFT_BOTTOM(9),
        /**
         * RIGHT_BOTTOM
         */
        RIGHT_BOTTOM(10);

        int v;

        /**
         * LEFT: 0
         * TOP: 2
         * RIGHT: 1
         * BOTTOM: 3
         */


        final static int[] PN = {1, -1};

        Anchor(int v) {
            this.v = v;
        }

        final static int COUNT = 4;

        public void move(RectF win, RectF frame, float dx, float dy) {
            float[] maxFrame = cohesion(win, CLIP_MARGIN);
            float[] minFrame = cohesion(frame, CLIP_FRAME_MIN);
            float[] theFrame = cohesion(frame, 0);

            float[] dxy = {dx, 0, dy};
            for (int i = 0; i < COUNT; i++) {
                if (((1 << i) & v) != 0) {

                    int pn = PN[i & 1];

                    theFrame[i] = pn * revise(pn * (theFrame[i] + dxy[i & 2]),
                            pn * maxFrame[i], pn * minFrame[i + PN[i & 1]]);
                }
            }

            frame.set(theFrame[0], theFrame[2], theFrame[1], theFrame[3]);
        }

        public static float revise(float v, float min, float max) {
            return Math.min(Math.max(v, min), max);
        }

        public static float[] cohesion(RectF win, float v) {
            return new float[]{
                    win.left + v, win.right - v,
                    win.top + v, win.bottom - v
            };
        }

        public static boolean isCohesionContains(RectF frame, float v, float x, float y) {
            return frame.left + v < x && frame.right - v > x
                    && frame.top + v < y && frame.bottom - v > y;
        }

        public static Anchor valueOf(int v) {
            Anchor[] values = values();
            for (Anchor anchor : values) {
                if (anchor.v == v) {
                    return anchor;
                }
            }
            return null;
        }
    }
}

File: imageedit/src/main/java/com/zhongjh/imageedit/core/clip/ImageClip.kt
package com.zhongjh.imageedit.core.clip


/**
 * 图像裁剪功能的基础接口，定义裁剪相关的常量和规范
 * @author zhongjh
 * @date 2025/09/01
 */
interface ImageClip {

    /**
     * 裁剪相关常量定义
     */
    companion object {
        /**
         * 裁剪区域与窗口边缘的边距，用于限制裁剪框最小内边距
         */
        const val CLIP_MARGIN = 60f

        /**
         * 裁剪框边角控制点的尺寸大小（边长）
         */
        const val CLIP_CORNER_SIZE = 48f

        /**
         * 裁剪区域的最小尺寸限制（宽/高的最小值）
         */
        const val CLIP_FRAME_MIN = CLIP_CORNER_SIZE * 3.14f

        /**
         * 裁剪框内部网格线的厚度
         */
        const val CLIP_THICKNESS_CELL = 3f

        /**
         * 裁剪框外边框的厚度
         */
        const val CLIP_THICKNESS_FRAME = 8f

        /**
         * 裁剪框边角线条的厚度
         */
        const val CLIP_THICKNESS_SEWING = 14f

        /**
         * 尺寸比例数组，用于计算网格线位置
         * 格式: {0f (起点), 1f (终点), 0.33f (1/3位置), 0.66f (2/3位置)}
         */
        val CLIP_SIZE_RATIO = floatArrayOf(0f, 1f, 0.33f, 0.66f)

        /**
         * 网格线坐标计算的位运算参数（内部使用）
         */
        const val CLIP_CELL_STRIDES = 0x7362DC98

        /**
         * 边角线条坐标计算的位运算参数（内部使用）
         */
        const val CLIP_CORNER_STRIDES = 0x0AAFF550

        /**
         * 边角线条的偏移步长数组，用于控制边角线条的延伸方向
         */
        val CLIP_CORNER_STEPS = floatArrayOf(0f, 3f, -3f)

        /**
         * 边角尺寸数组，用于计算边角线条的长度
         * 格式: {0f (无偏移), CLIP_CORNER_SIZE (正向偏移), -CLIP_CORNER_SIZE (负向偏移)}
         */
        val CLIP_CORNER_SIZES = floatArrayOf(0f, CLIP_CORNER_SIZE, -CLIP_CORNER_SIZE)

        /**
         * 边角线条的配置数组，存储每个边角线条的计算参数（内部使用）
         */
        val CLIP_CORNERS = byteArrayOf(
            0x8, 0x8, 0x9, 0x8,
            0x6, 0x8, 0x4, 0x8,
            0x4, 0x8, 0x4, 0x1,
            0x4, 0xA, 0x4, 0x8,
            0x4, 0x4, 0x6, 0x4,
            0x9, 0x4, 0x8, 0x4,
            0x8, 0x4, 0x8, 0x6,
            0x8, 0x9, 0x8, 0x8
        )
    }
}